### PR TITLE
Github Actions CI: streamline build, export all artifacts required by testing, improve testing

### PR DESCRIPTION
**Pull request by t-lo**
This PR streamlines Github actions CI builds and makes build steps re-use the build container, saving some build time.

It also adds previously missing artifacts required by the tests. It makes artifact upload more granular, so build logs, images, dev container, and torcx packages can be downloaded individually. It exports torcx tarball and binary packages for successive re-use in the kola tests.

The change also updates the github actions kola test runner workflow to use the new, separated artifacts produced by ci.yaml.

Further, it adds a fix for the devcontainer tests. Devcontainer and bin packages used in the devcontainer tests are now served from a local temporary web server.

The change also adds the qemu_update test and provides the respective update payload.

Also, the tests now use a local torcx_manifest.json produced by ci.yaml, which points to a torcx tarball also served by the local temporary web server.

Lastly, the change adds Markdown export capabilities to `tapfile_helper_lib.sh` to create a test run summary suitable for ingestion by Github. Test result reports for each run are posted to PRs as comments.

## How to use

Go to the ["Run build" GH action](https://github.com/flatcar/scripts/actions/workflows/ci.yaml), click on the "Run workflow" drop-down,  select branch "t-lo/speed-up-ci-build", and click on the green "Run workflow" button.

## Testing done

Ran the workflow.

## Cherry-pick and backport

This PR should be cherry-picked into all active maintenance branches (including LTS) as it would allow us to run automated builds and tests for auto-generated PRs (kernel, systemd, package refreshes).

### CODE REVIEW

File: .github/workflows/run-kola-tests.yaml
Comment (pothos) [inline .github/workflows/run-kola-tests.yaml:166]:
```suggestion
          IMG_ARCHIVE=$(readlink -f images/**/flatcar_production_image.bin.bz2)
```
Or is this something else?

File: .github/workflows/run-kola-tests.yaml
Comment (pothos) [inline .github/workflows/run-kola-tests.yaml:155]:
```suggestion
          tar -C ${DEVCONTAINER_WEBROOT} -xf binpkgs.tar.gz
```
Or `xvf` but maybe we don't need the full output?

File: .github/workflows/run-kola-tests.yaml
Comment (t-lo) [reply .github/workflows/run-kola-tests.yaml]: That's a leftover which isn't actually needed anymore (no subdirectories anymore for image artifacts). Will remove, please pardon the confusion.

File: .github/workflows/run-kola-tests.yaml
Comment (t-lo) [reply .github/workflows/run-kola-tests.yaml]: Yeah, that one just bit me in the actions run I triggered before signing off yesterday. Needs to be `-xzvf` (the `v` is indeed a good idea).

**Comment by pothos** [comment]
In https://github.com/flatcar/scripts/pull/697/files I also added a new environment variable to the kola workflow. It will probably be a git conflict.

**Comment by t-lo** [comment]
> In https://github.com/flatcar/scripts/pull/697/files I also added a new environment variable to the kola workflow. It will probably be a git conflict.

Good catch @pothos, thank you! This looks like an easy merge though. I'll rebase on main when your PR was merged.

**Comment by t-lo** [comment]
This should take care of the devcontainer test breakage: https://github.com/flatcar/flatcar-dev-util/pull/14 .

**Comment by t-lo** [comment]
Rebased on latest main.

**Comment by t-lo** [comment]
Included the emerge-gitclone fix from latest coreos-overlay:main.

Let's see if we can trigger a rebuild+retest by re-requesting reviews. This is one of the trigger actions for `ci.yaml` and is supposed to work...

**Comment by t-lo** [comment]
Test results artifacts are 10GB each, this seems excessive. Taking a look.

File: .github/workflows/ci.yaml
Comment (pothos) [inline .github/workflows/ci.yaml:83]: I think we should also do this in the kola workflow for consistency - otherwise it's confusing if one works on the merged state and the other not.

**Review by pothos — COMMENTED** [review]
Looks good!
The `coreos-overlay: bump ref to include emerge-gitclone fix` commit needs to go before merging, as we would not want to backport this submodule reference update but have to do backports to the `flatcar-MAJOR` branches of coreos-overlay.

File: .github/workflows/ci.yaml
Comment (t-lo) [reply .github/workflows/ci.yaml:83]: Oh right! Good catch.

**Comment by t-lo** [comment]
> Looks good! The `coreos-overlay: bump ref to include emerge-gitclone fix` commit needs to go before merging, as we would not want to backport this submodule reference update but have to do backports to the `flatcar-MAJOR` branches of coreos-overlay.

Good point. I'll split this out into a separate PR then.

**Comment by t-lo** [comment]
A few things are still open.

**TODO**:
- [ ] Split coreos-overlay bump into a separate PR and create respective PRs for flatcar-3033, flatcar-3549, flatcar-3510, flatcar-3374.
- [ ] Use `${{ github.event.pull_request.head.sha }}` for PR test runs in `run-kola-tests.yaml` in accordance with `ci.yaml`
- [ ] Don't upload 10GB of OS image artifacts generated by the qemu update test.
- [ ] For some reason the merged test runs TAP report is not present after testing concluded. Figure out why and archive the merged tests TAP report.

**Comment by t-lo** [comment]
Manually triggered a new test run here: https://github.com/flatcar/scripts/actions/runs/4541175015

**Comment by t-lo** [comment]
Everything should be in place now. Re-triggering a build/test run from scratch (by re-requesting review from pothos :) )

**Review by pothos — APPROVED** [review]

**Review by gabriel-samfira — APPROVED** [review]
This all looks reasonable. LGTM.

File: .github/workflows/run-kola-tests.yaml
Comment (gabriel-samfira) [inline .github/workflows/run-kola-tests.yaml:261]: These files need to have the first line one of the following patters:

https://github.com/test-summary/action/blob/b266d9a22ea23552e00a4ee4276543880f5f686e/src/test_parser.ts#L299-L303

Otherwise it tries to parse them as XML, and you get the ```Error: Non-whitespace before first tag.``` message.

Comment (t-lo) [reply .github/workflows/run-kola-tests.yaml:261]: It appears that we need to prepend a magic "TAP version 13" line. There are a [few examples in their unit tests](https://github.com/test-summary/action/tree/b266d9a22ea23552e00a4ee4276543880f5f686e/test/resources/tap) that suggest the format we emit should otherwise work. I'll patch and re-run.

Comment (t-lo) [reply .github/workflows/run-kola-tests.yaml]: Oh wow that TAP parser is brittle. It insists on 2 spaces and who knows what else for embedded YAML. This makes me sad. TAP is a loosely defined format and it is meant to be interpreted loosely, not strictly.

I'll step away from that action and will investigate other options.

**Comment by t-lo** [comment]
I was required to change quite a bit to make test results reporting usable for GitHub. Adding @krnowak as a reviewer because of his awesome bash skills and re-requested review from Kai.

**Comment by krnowak** [comment]
I'll have a look tomorrow.

**Review by pothos — APPROVED** [review]

**Comment by pothos** [comment]
One thing to improve is that the amd64 and arm64 status should not be posted to the same comment. Now one has to klick `edited` to see the first one of the status reports.

**Comment by t-lo** [comment]
> One thing to improve is that the amd64 and arm64 status should not be posted to the same comment. Now one has to klick `edited` to see the first one of the status reports.

I'm a little confused it did that - but then I'm still learning the ropes with the add-pr-comment action.

Something we could do on our end to improve this is to merge both amd64 and arm64 into one report, using `tapfile_helper_lib.sh`. We could pass all TAP files to a separate merger action and use `qemu_uefi (amd64)` and `qemu_uefi (arm64)` as "vendors". This would have us end up with a report that includes all runs and would not require further changes to `tapfile_helper_lib.sh`. I'll have a look later.

I'll wait for Krzesimir's review (because I touched a lot of bash) and merge if I get his LGTM.

File: .github/workflows/run-kola-tests.yaml
Comment (krnowak) [inline .github/workflows/run-kola-tests.yaml:276]: Indentation seems to be off.

**Comment by t-lo** [comment]
@pothos I've added test report merging in a separate job; we will now generate a single test report only, with amd64 and arm64 tests merged in (both qemu_uefi and qemu_update). I've kicked off a test to provide an example of a merged report, please have a look after it concluded: https://github.com/flatcar/scripts/actions/runs/4563204306

File: ci-automation/tapfile_helper_lib.sh
Comment (krnowak) [inline ci-automation/tapfile_helper_lib.sh:255]: `finish_test_report` vfunc seems to be unnecessary, as both `tap` and `md` just return `true`. Are there plans for some other format, where `finish_test_report` would be useful?

Comment (krnowak) [reply ci-automation/tapfile_helper_lib.sh:255]: Well, I guess it will be useful when implementing json format. Nevermind.

**Review by krnowak — APPROVED** [review]

File: ci-automation/tapfile_helper_lib.sh
Comment (t-lo) [reply ci-automation/tapfile_helper_lib.sh:255]: Correct! @gabriel-samfira mentioned a desire for JSON at some point.

File: .github/workflows/run-kola-tests.yaml
Comment (t-lo) [reply .github/workflows/run-kola-tests.yaml]: Good catch; fixed.

**Comment by t-lo** [comment]
Merged and cherry-picked to flatcar-3033, flatcar-3549, flatcar-3510, flatcar-3374.

**Pull request merged**
